src/dtls13.c

```diff
@@ -0,0 +1,1000 @@
+            break;
+
+        ret = DoTls13HandShakeMsgType(ssl, msg->msg, &idx, msg->type,
+                                      msg->sz, msg->sz);
+        if (ret != 0)
+            break;
+
+        /* TODO: update the retransmission logic with new message processed */
+
+        /* FIXME: move in a better place. handle overflow? */
+        ssl->keys.dtls_expected_peer_handshake_number++;
+
+        ssl->dtls_rx_msg_list = msg->next;
+        DtlsMsgDelete(msg, ssl->heap);
+        msg = ssl->dtls_rx_msg_list;
+        ssl->dtls_rx_msg_list_sz--;
+    }
+
+    WOLFSSL_LEAVE("dtls13_process_buffered_messages()", ret);
+
+    return ret;
+}
+
+static int dtls13_next_message_complete(WOLFSSL *ssl) {
+    return ssl->dtls_rx_msg_list != NULL &&
+        ssl->dtls_rx_msg_list->fragSz >= ssl->dtls_rx_msg_list->sz;
+}
+
+static inline int frag_is_in_output_buffer(WOLFSSL *ssl, const byte *frag)
+{
+    const byte *output_buffer = ssl->buffers.outputBuffer.buffer;
+    word16 output_buffer_size = ssl->buffers.outputBuffer.bufferSize;
+
+    return frag >= output_buffer && frag < output_buffer + output_buffer_size;
+}
+
+static int dtls13_send_frag_from_buffer(WOLFSSL *ssl,
+                                        byte *output, word16 length)
+{
+    byte *buf;
+    int ret;
+
+    ret = CheckAvailableSize(ssl, length);
+    if(ret != 0)
+        return ret;
+
+    buf = ssl->buffers.outputBuffer.buffer +
+        ssl->buffers.outputBuffer.length;
+
+    XMEMCPY(buf, output, length);
+
+    ssl->buffers.outputBuffer.length += length;
+
+    return SendBuffered(ssl);
+}
+
+static int dtls13_send_one_fragment(WOLFSSL *ssl, byte *output,
+                                    word16 output_size,
+                                    word16 length,
+                                    enum HandShakeType handshake_type,
+                                    int hash_output)
+{
+    word16 record_header_length;
+    word16 record_length;
+    byte is_protected;
+    int send_length;
+    byte *msg;
+    int ret;
+
+
+    if (output_size < length)
+        return BUFFER_ERROR;
+
+    is_protected = dtls13_type_is_encrypted(handshake_type);
+    record_header_length = dtls13_get_rl_header_length(ssl, is_protected);
+
+    if (length <= record_header_length)
+            return BUFFER_ERROR;
+
+    record_length = length - record_header_length;
+
+    /* plain text (ClientHello/ServerHello/ACK) */
+    if (!is_protected) {
+
+        ret = dtls13_rl_add_plaintext_header(ssl,
+                                             output, handshake, record_length);
+        if (ret != 0)
+            return ret;
+
+        msg = output + DTLS_RECORD_HEADER_SZ;
+
+    } else {
+        /* protected message  */
+        msg = output + record_header_length;
+
+        if (length <= record_header_length)
+            return BUFFER_ERROR;
+
+        /* FIXME: ignore async ok flag for now */
+        send_length = BuildTls13Message(ssl, output,
```

Comment:
Formatting
Similar to func decls, if it is going to squeeze right then put parameters on next line 4 more spaces in then original line.